File: Homework/Hw7/Hw7.java

import java.util.List;
import java.util.Arrays;
import java.util.PriorityQueue;

public class Hw7 {

    public static int minMeetingRooms(int[][] intervals) {
        if (intervals == null || intervals.length == 0) {
            return 0;
        }
        
        Arrays.sort(intervals, (a, b) -> Integer.compare(a[0], b[0]));
        
        PriorityQueue<Integer> allocator = new PriorityQueue<>(intervals.length, (a, b) -> Integer.compare(a, b));
        
        allocator.add(intervals[0][1]);
        
        for (int i = 1; i < intervals.length; i++) {
            if (intervals[i][0] >= allocator.peek()) {
                allocator.poll();
            }
            allocator.add(intervals[i][1]);
        }
        
        return allocator.size();
    }
    
    public static void main(String[] args) {
        int[][] intervals = {{0, 30}, {5, 10}, {15, 20}};
        System.out.println("Minimum number of meeting rooms required: " + minMeetingRooms(intervals));
        
        int[][] intervals2 = {{7, 10}, {2, 4}};
        System.out.println("Minimum number of meeting rooms required: " + minMeetingRooms(intervals2));
    }
}

File: Homework/Hw7/Hw7.py
from typing import List
import heapq

class Hw7:
    def minMeetingRooms(self, intervals: List[List[int]]) -> int:
        if not intervals:
            return 0

        # Start by sorting the intervals by the start time.
        intervals.sort(key=lambda x: x[0])

        # Initialize a heap.
        heap = []

        # Add the first meeting. We have to give a new room to the first meeting.
        heapq.heappush(heap, intervals[0][1])

        # For all the remaining meeting intervals...
        for interval in intervals[1:]:
            # If the room due to free up the earliest is free, assign that room to this meeting.
            if heap[0] <= interval[0]:
                heapq.heappop(heap)

            # If a new room is to be assigned, then also we add to the heap.
            # If an old room is allocated, then also we have to add to the heap with updated end time.
            heapq.heappush(heap, interval[1])

        # The size of the heap tells us the minimum rooms required for all the meetings.
        return len(heap)

hw7 = Hw7()
intervals = [[0, 30], [5, 10], [15, 20]]
print(hw7.minMeetingRooms(intervals))  # Output: 2
